### ISSUE TITLE
CLI command for adding patches

### ISSUE THREAD

**Issue by danepowell** [cweagans/composer-patches#81]
In the same way that Composer includes a "require" command, so that you can run `composer require cweagans/composer-patches` instead of editing `composer.json` to insert the dependency, it would be really nice if composer-patches provided a way to patch packages from the command line.

I could see this either taking the form of a `patch` command, or an additional `patch` parameter to the `require` command (if Composer lets you modify built-in commands... not sure).

**Comment by danepowell** [comment]
For anyone looking for a workaround, I can't really find one. I was hoping `composer config` would work but it doesn't allow setting an array below [three levels deep](https://getcomposer.org/doc/03-cli.md#modifying-extra-values). So you can't add a new patch without deleting all of the existing ones.

**Comment by weitzman** [comment]
I would really use this as well. And thanks @danepowell for saying that the workaround I had in mind doesn't work either. Have you found anything besides loading combo.json and writing it back to disk? Ping @grasmash as well.

**Comment by grasmash** [comment]
This package could provide a static method that would read the composer.json file, modify, and write it to disk. That method can then be placed in a script in your composer.json:
```
  "scripts": {
    "add-patch": "Cweagens\\ComposerPatches\\Plugin::add-patch"
  }
```

Allowing you to call `composer add-patch [patch.patch]`.
```

**Comment by cweagans** [comment]
Plugins can provide commands, I think. No need to define a separate script in your composer.json.

**Comment by alisonjo2786** [comment]
Yes: https://getcomposer.org/doc/articles/plugins.md#command-provider

And I see references to command-y things here...
https://github.com/cweagans/composer-patches/blob/master/src/Patches.php#L354

But I definitely don't understand how I'm meant to use it/them!

**Comment by cweagans** [comment]
The steps are basically as follows:

* Implement `Composer\Plugin\Capability\CommandProvider` in `cweagans\Composer\Patches` - this should return an array of Objects.
* Create a new class that extends `Composer\Command\BaseCommand`. `configure()` should set the name of the command (i.e. `patch-add` or whatever - see example you linked for the exact syntax) and `execute()` should actually do the work of loading composer.json (or the configured patches file) and adding the patch.

If you start working on this and get stuck, push your code somewhere and let me know. Happy to help as needed.

**Comment by axel-rutz** [comment]
I think this is a very useful case of a generic CLI json modifier. I remember there was a drupal-console generic yaml modifyer.

Would really love to see both (the generic one and the patches one, maybe just depending).

**Comment by cweagans** [comment]
This is on the roadmap for 2.0.0 (see #93)

**Comment by szeidler** [comment]
Nice to see, that the CLI is on the Roadmap for 2.x.

Recently I started an initial composer patches CLI plugin for the 1.x branch at https://github.com/szeidler/composer-patches-cli , to test a bit, what is possible and what not.

We used it for adding the recent Drupalgeddon security patches in an automated way.

**Comment by cweagans** [comment]
I saw that. :) I'll probably end up using a fair amount of your code -- there's a lot of good stuff there.

**Comment by axel-rutz** [comment]
Note: The mentioned szeidler/composer-patches-cli **only** works for adding patches via extra/patches-file. It does **not** work for patches added via extra/patches.

**Comment by christopher-hopper** [comment]
Would it be possible, as part of this, to allow a little bit of optional metadata to be stored with a patch. Thinking like:

- patch source (ie. github, drupal, bitbucket, gist) 
- patch issue id (ie. 2493183, ABC-123)
- patch issue URL (ie. https://www.drupal.org/node/2493183, https://github.com/vendor/repo/issues/81 )

This would be helpful so that add-ons / plug-ins could be written to support management of patches by calling issue queue apis, to check on the status of an issue associated with a patch.

**Comment by frederickjh** [comment]
@christopher-hopper, @cweagans stated above that he was planning to use a lot of  @szeidler's code from [szeidler/composer-patches-cli](https://github.com/szeidler/composer-patches-cli). This repository includes the functionality that you are asking for.

**Comment by cweagans** [comment]
@christopher-hopper Please open an issue for that. I'm not opposed to adding additional optional metadata fields for `Patch` objects in the 2.x releases.

**Comment by cweagans** [comment]
I've made a note to handle this in `main`.

**Comment by MurzNN** [comment]
So, is this feature implemented in the core functionality, or not yet? I see that it is closed as completed on Feb 8, 2023, but I can't find any documentation about how to use the CLI.

**Comment by cweagans** [comment]
I haven't added it yet. Feel free!
